src/tike/ptycho/exitwave.py

```diff
@@ -0,0 +1,180 @@
+    I_m             : ND array ( # scan positions, # rows, # columns ), measured diffraction intensity
+    I_e             : ND array ( # scan positions, # rows, # columns ), calculated diffraction intensity 
+    measured_pixels : 2D binary array ( # rows, # columns ), the regions on the detector where we have defined measurements
+    step_length     : 2D array ( # shared probe modes, # scan positions), the steplength initializations
+    weight_avg      : float ( 0.0 <= weight_avg <= 1.0  ), the weight we use when computing a weighted average
+    """
+
+    if measured_pixels.size == 0:
+        measured_pixels = 1
+
+    xi_abs_Psi2 = xi * abs2_Psi
+
+    for _ in range(0, 2):
+
+        xi_alpha_minus_one = xi * step_length[..., None, None] - 1
+
+        numer = I_m * xi_alpha_minus_one
+        denom = abs2_Psi * cp.square(xi_alpha_minus_one) + I_e - abs2_Psi
+        numer = cp.sum(measured_pixels * xi_abs_Psi2 * (1 + numer / denom),
+                       axis=(-1, -2))
+
+        denom = cp.sum(measured_pixels * cp.square(xi) * abs2_Psi,
+                       axis=(-1, -2))
+
+        step_length = step_length * (1 - weight_avg) + (numer /
+                                                        denom) * weight_avg
+
+    return step_length
+
+
+def poisson_steplength_dominant_mode(
+    xi,
+    I_e,
+    I_m,
+    measured_pixels,
+    step_length,
+    weight_avg,
+):
+    """Compute the optimal steplength for each exitwave mode using only the dominant mode.
+
+    Parameters
+    ----------
+    xi              : ND array ( # scan positions, # rows, # columns ), xi = 1 - I_m / I_e
+    I_m             : ND array ( # scan positions, # rows, # columns ), measured diffraction intensity
+    I_e             : ND array ( # scan positions, # rows, # columns ), calculated diffraction intensity 
+    measured_pixels : 2D binary array ( # rows, # columns ), the regions on the detector where we have defined measurements
+    step_length     : 2D array ( # shared probe modes, # scan positions), the steplength initializations
+    weight_avg      : float ( 0.0 <= weight_avg <= 1.0  ), the weight we use when computing a weighted average
+    """
+
+    if measured_pixels.size == 0:
+        measured_pixels = 1
+
+    denom = measured_pixels * I_e * cp.square(xi)
+
+    sum_denom = cp.sum(denom, axis=(-1, -2))
+
+    for _ in range(0, 2):
+
+        nom = measured_pixels * xi * (I_e - I_m /
+                                      (1 - step_length[..., None, None] * xi))
+
+        nom_over_denom = cp.sum(nom, axis=(-1, -2)) / sum_denom
+
+        step_length = (1 -
+                       weight_avg) * step_length + weight_avg * nom_over_denom
+
+        step_length = cp.abs(cp.fmax(cp.fmin(step_length, 1), 0))
+
+    step_length += 1e-2 * tike.random.randomizer_cp.standard_normal(
+        size=step_length.shape,
+        dtype=tike.precision.floating,
+    )
```

Comment:
Does this randomness improve the result? I'm not sure it's performant to compute this many random numbers every epoch.

Reply:
Please remove the randomness from the steplength. I think there's other problems such as assuming that the steplength should always be on same order as 10^-2?

Reply:
see new commit